fastchat/serve/gradio_block_arena_anony.py

```diff
@@ -191,52 +200,113 @@ def share_click(state0, state1, model_selector0, model_selector1, request: gr.Re
+        "gpt-4": {"claude-2"},
+        "gpt-4-turbo": {"gpt-4", "gpt-3.5-turbo"},
+        # "gpt-4": {"llama-2-70b-chat"},
+        "gpt-3.5-turbo": {"claude-instant-1", "gpt-4", "claude-2"},
+        # "gpt-3.5-turbo": {"llama-2-70b-chat"},
+        "claude-2": {"gpt-4", "gpt-3.5-turbo", "claude-1"},
+        "claude-1": {"claude-2", "gpt-4", "gpt-3.5-turbo"},
+        "claude-instant-1": {"gpt-3.5-turbo", "claude-2"},
+        "deluxe-chat-v1.1": {"gpt-4"},
+        "openchat-3.5": {"gpt-3.5-turbo", "llama-2-70b-chat", "zephyr-7b-beta"},
+        "qwen-14b-chat": {"vicuna-13b", "llama-2-13b-chat", "llama-2-70b-chat"},
+        "zephyr-7b-alpha": {"mistral-7b-instruct", "llama-2-13b-chat"},
+        "zephyr-7b-beta": {
+            "mistral-7b-instruct",
+            "llama-2-13b-chat",
+            "llama-2-7b-chat",
+            "wizardlm-13b",
+        },
+        "llama-2-70b-chat": {"gpt-3.5-turbo", "vicuna-33b", "claude-instant-1"},
+        "llama-2-13b-chat": {"mistral-7b-instruct", "vicuna-13b", "llama-2-70b-chat"},
+        "llama-2-7b-chat": {"mistral-7b-instruct", "vicuna-7b", "llama-2-13b-chat"},
+        "mistral-7b-instruct": {
+            "llama-2-7b-chat",
+            "llama-2-13b-chat",
+            "llama-2-70b-chat",
+        },
+        "vicuna-33b": {"llama-2-70b-chat", "gpt-3.5-turbo", "claude-instant-1"},
+        "vicuna-13b": {"llama-2-13b-chat", "llama-2-70b-chat"},
+        "vicuna-7b": {"llama-2-7b-chat", "mistral-7b-instruct", "llama-2-13b-chat"},
+        "wizardlm-70b": {"gpt-3.5-turbo", "vicuna-33b", "claude-instant-1"},
+        "palm-2": {"llama-2-13b-chat", "gpt-3.5-turbo"},
+    }
+    model_weights = []
+    for model in models:
+        weight = get_sample_weight(model)
+        model_weights.append(weight)
+    total_weight = np.sum(model_weights)
+    model_weights = model_weights / total_weight
+    chosen_idx = np.random.choice(len(models), p=model_weights)
+    chosen_model = models[chosen_idx]
+
+    rival_models = []
+    rival_weights = []
+    for model in models:
+        if model == chosen_model:
+            continue
+        weight = get_sample_weight(model)
+        if weight != 0 and chosen_model in targets and model in targets[chosen_model]:
+            # boost to 50% chance
+            weight = total_weight / len(targets[chosen_model])
+        rival_models.append(model)
+        rival_weights.append(weight)
+    # for p, w in zip(rival_models, rival_weights):
+    #     print(p, w)
+    rival_weights = rival_weights / np.sum(rival_weights)
+    rival_idx = np.random.choice(len(rival_models), p=rival_weights)
+    rival_model = rival_models[rival_idx]
+
+    swap = np.random.randint(2)
+    if swap == 0:
+        return chosen_model, rival_model
+    else:
+        return rival_model, chosen_model
 
 
 def add_text(
     state0, state1, model_selector0, model_selector1, text, request: gr.Request
 ):
-    ip = request.client.host
+    is_cf = "cf-connecting-ip" in request.headers
```

Comment:
delete this. It appears multiple times in this file

Reply:
great catch!